### PR TITLE
Primary documentation for Great Deluge algorhitm

### PR DESCRIPTION
**Pull request by BorisKocian**
The main documentation for the Great Deluge algorithm.

### CODE REVIEW

**Comment by kie-ci10** [comment]
Can one of the admins verify this PR? Comment with 'ok to test' to start the build.

**Comment by ge0ffrey** [comment]
ok to test

File: optaplanner-docs/src/main/asciidoc/LocalSearch/LocalSearch-chapter.adoc [hunk 1]
```diff
@@ -476,6 +476,46 @@ Use a lower tabu size than in a pure Tabu Search configuration.
...
 
+[[greatDeluge]]
+== Great Deluge
+
+
+[[greatDelugeAlgorithm]]
+=== Algorithm Description
+
+Great Deluge algorithm is very similar to the Simulated Annealing algorithm, it evaluates only a few moves per steps,
+so it steps quickly. The first accepted move is the winning step. A move is accepted only if it does not decrease
```
Comment (ge0ffrey) [inline optaplanner-docs/src/main/asciidoc/LocalSearch/LocalSearch-chapter.adoc]: "decrease" -> "is lower than"

Those that aren't lower than the water level, don't decrease it either.

File: optaplanner-docs/src/main/asciidoc/LocalSearch/LocalSearch-chapter.adoc [hunk 2]
```diff
@@ -476,6 +476,46 @@ Use a lower tabu size than in a pure Tabu Search configuration.
...
+the score value (water level) that we are working with. It means Great Deluge is deterministic and opposite
+of Simulated Annealing has no randomization in it. The water level is increased after every step either about the fixed value
+or by percentual value.
+A gradual increase in water level gives Great Deluge more time to escape from local maxima.
+
+
+[[greatDelugeConfiguration]]
+=== Configuration
+
+`GreatDelugeStartingWaterLevel` can be set as a starting water level but is recommended to do not do it,
```
Comment (ge0ffrey) [inline optaplanner-docs/src/main/asciidoc/LocalSearch/LocalSearch-chapter.adoc]: Start with the required parameters and go to the more-and-more optional parameters in order. Show the xml first.

File: optaplanner-docs/src/main/asciidoc/LocalSearch/LocalSearch-chapter.adoc [hunk 3]
```diff
@@ -476,6 +476,46 @@ Use a lower tabu size than in a pure Tabu Search configuration.
...
+so the algorithm takes as starting value best score from construction heuristic.
+Either `greatDelugeRainSpeedScore` or `greatDelugeRainSpeedRatio` have to be set.
+If `greatDelugeRainSpeedScore` is set, the water level is increased by a constant value but is recommended to use
+`greatDelugeRainSpeedRatio` when the water level is increased by percentual value, so there is no need to know the size of the problem or value of a scoring function.
+The good starting value of `greatDelugeRainSpeedRatio` may be `0.99999995`.
```
Comment (ge0ffrey) [inline optaplanner-docs/src/main/asciidoc/LocalSearch/LocalSearch-chapter.adoc]: grammar: "The good ..." -> "A good ..."

I wouldn't mention the starting value at all - it's already the default if nothing is filled in anyway.

File: optaplanner-docs/src/main/asciidoc/LocalSearch/LocalSearch-chapter.adoc [hunk 4]
```diff
@@ -476,6 +476,46 @@ Use a lower tabu size than in a pure Tabu Search configuration.
   </localSearch>
 ----
 
+[[greatDeluge]]
+== Great Deluge
```
Comment (ge0ffrey) [inline optaplanner-docs/src/main/asciidoc/LocalSearch/LocalSearch-chapter.adoc]: Great Deluge must come after Tabu Search, Late Acceptance and Simulated Annealing. I am ok to having it before Step Counting Hill Climbing. It must first earn it's spur before it's promoted to a top position :)

File: optaplanner-docs/src/main/asciidoc/LocalSearch/LocalSearch-chapter.adoc [hunk 5]
```diff
@@ -476,6 +476,46 @@ Use a lower tabu size than in a pure Tabu Search configuration.
...
+
+
+[[greatDelugeAlgorithm]]
+=== Algorithm Description
+
+Great Deluge algorithm is very similar to the Simulated Annealing algorithm, it evaluates only a few moves per steps,
```
Comment (ge0ffrey) [inline optaplanner-docs/src/main/asciidoc/LocalSearch/LocalSearch-chapter.adoc]: Great Deluge is far more similar to Late Acceptance than it is to Simulated Annealing. But yes, all those 3 are fast-stepping algorithms.

File: optaplanner-docs/src/main/asciidoc/LocalSearch/LocalSearch-chapter.adoc [hunk 6]
```diff
@@ -476,6 +476,46 @@ Use a lower tabu size than in a pure Tabu Search configuration.
...
+so it steps quickly. The first accepted move is the winning step. A move is accepted only if it does not decrease
+the score value (water level) that we are working with. It means Great Deluge is deterministic and opposite
+of Simulated Annealing has no randomization in it. The water level is increased after every step either about the fixed value
+or by percentual value.
+A gradual increase in water level gives Great Deluge more time to escape from local maxima.
+
+
+[[greatDelugeConfiguration]]
+=== Configuration
```
Comment (ge0ffrey) [inline optaplanner-docs/src/main/asciidoc/LocalSearch/LocalSearch-chapter.adoc]: Just like Tabu Search configuration, have a "Simplest configuration" and an "Advanced configuration". Make it very consistent with the other sections already there, so it's easier for the reader to understand it.

https://docs.optaplanner.org/latestFinal/optaplanner-docs/html_single/#tabuSearchConfiguration

Comment (ge0ffrey) [reply optaplanner-docs/src/main/asciidoc/LocalSearch/LocalSearch-chapter.adoc]: Note: Simulated annealing is missing a "simplest configuration". That's a TODO for us.

File: optaplanner-docs/src/main/asciidoc/LocalSearch/LocalSearch-chapter.adoc (hunk 4 above)
Comment (ge0ffrey) [reply optaplanner-docs/src/main/asciidoc/LocalSearch/LocalSearch-chapter.adoc]: Add a row in the table in this section:
https://docs.optaplanner.org/latestFinal/optaplanner-docs/html_single/#optimizationAlgorithmsOverview

File: optaplanner-docs/src/main/asciidoc/LocalSearch/LocalSearch-chapter.adoc (hunk 5 above)
Comment (ge0ffrey) [reply optaplanner-docs/src/main/asciidoc/LocalSearch/LocalSearch-chapter.adoc]: When mentioning another algorithm, link it. So <<simulatedAnnealing,Simulated Annealing>>

File: optaplanner-docs/src/main/asciidoc/LocalSearch/LocalSearch-chapter.adoc [hunk 7]
```diff
@@ -542,6 +542,70 @@ Use a lower tabu size than in a pure Tabu Search configuration.
...
+
+If `greatDelugeWaterLevelIncrementScore` is set, the water level is increased by a constant value.
+
+[source,xml,options="nowrap"]
+----
+<acceptor>
+  <greatDelugeWaterLevelIncrementScore>100</greatDelugeWaterLevelIncrementScore>
+</acceptor>
+----
+To avoid hard coding the rain speed score, configure a `greatDelugeWaterLevelIncrementRatio` (recommended) when the water level is increased by percentual value, so there is no need to know the size of the problem or value of a scoring function.
```
Comment (triceo) [inline optaplanner-docs/src/main/asciidoc/LocalSearch/LocalSearch-chapter.adoc]: You're speaking of "rain speed", while you should newly be speaking of "water level increment", I think.

File: optaplanner-docs/src/main/asciidoc/LocalSearch/LocalSearch-chapter.adoc [hunk 8]
```diff
@@ -542,6 +542,70 @@ Use a lower tabu size than in a pure Tabu Search configuration.
...
+
+[source,xml,options="nowrap"]
+----
+<acceptor>
+  <greatDelugeWaterLevelIncrementRatio>100</greatDelugeWaterLevelIncrementRatio>
+</acceptor>
+----
+
+Also `greatDelugeInitialWaterLevel` can be set as a starting water level but is recommended to do not do it,
```
Comment (triceo) [inline optaplanner-docs/src/main/asciidoc/LocalSearch/LocalSearch-chapter.adoc]: `is recommended not to do it`

File: optaplanner-docs/src/main/asciidoc/LocalSearch/LocalSearch-chapter.adoc [hunk 9]
```diff
@@ -542,6 +542,70 @@ Use a lower tabu size than in a pure Tabu Search configuration.
...
+so the algorithm takes as starting value best score from construction heuristic.
```
Comment (triceo) [inline optaplanner-docs/src/main/asciidoc/LocalSearch/LocalSearch-chapter.adoc]: `the best score`

Comment (triceo) [reply optaplanner-docs/src/main/asciidoc/LocalSearch/LocalSearch-chapter.adoc]: `from the construction heuristic`

File: optaplanner-docs/src/main/asciidoc/LocalSearch/LocalSearch-chapter.adoc [hunk 10]
```diff
@@ -542,6 +542,70 @@ Use a lower tabu size than in a pure Tabu Search configuration.
...
+Use the Benchmarker to fine tweak your configuration.
```
Comment (triceo) [inline optaplanner-docs/src/main/asciidoc/LocalSearch/LocalSearch-chapter.adoc]: `fine-tune`

File: optaplanner-docs/src/main/asciidoc/LocalSearch/LocalSearch-chapter.adoc [hunk 11]
```diff
@@ -542,6 +542,70 @@ Use a lower tabu size than in a pure Tabu Search configuration.
...
+    <acceptor>
+      <greatDelugeInitialWaterLevel>20hard/100soft</greatDelugeInitialWaterLevel>
+      <greatDelugeWaterLevelIncrementRatio>0.00000005</greatDelugeWaterLevelIncrementRatio>
+    </acceptor>
+    <forager>
+      <acceptedCountLimit>4</acceptedCountLimit>
+    </forager>
+  </localSearch>
+----
+Planner implements two water level increment options:
```
Comment (triceo) [inline optaplanner-docs/src/main/asciidoc/LocalSearch/LocalSearch-chapter.adoc]: `OptaPlanner`?

Comment (triceo) [reply optaplanner-docs/src/main/asciidoc/LocalSearch/LocalSearch-chapter.adoc]: Also, @ge0ffrey, I'm not sure how we handle this in productized documentation. Isn't there some product name placeholder?

File: optaplanner-docs/src/main/asciidoc/LocalSearch/LocalSearch-chapter.adoc [hunk 12]
```diff
@@ -542,6 +542,70 @@ Use a lower tabu size than in a pure Tabu Search configuration.
...
+<acceptor>
+  <greatDelugeWaterLevelIncrementScore>100</greatDelugeWaterLevelIncrementScore>
+</acceptor>
+----
+To avoid hard coding the water level increment, configure a `greatDelugeWaterLevelIncrementRatio` (recommended) when the water level is increased by percentual value, so there is no need to know the size of the problem or value of a scoring function.
+
+[source,xml,options="nowrap"]
+----
+<acceptor>
+  <greatDelugeWaterLevelIncrementRatio>100</greatDelugeWaterLevelIncrementRatio>
```
Comment (triceo) [inline optaplanner-docs/src/main/asciidoc/LocalSearch/LocalSearch-chapter.adoc]: I'm pretty sure 100 is a wrong value here.

File: optaplanner-docs/src/main/asciidoc/LocalSearch/LocalSearch-chapter.adoc [hunk 13]
```diff
@@ -542,6 +542,70 @@ Use a lower tabu size than in a pure Tabu Search configuration.
...
+[source,xml,options="nowrap"]
+----
+  <localSearch>
+    ...
+    <acceptor>
+      <greatDelugeInitialWaterLevel>20hard/100soft</greatDelugeInitialWaterLevel>
+      <greatDelugeWaterLevelIncrementRatio>0.00000005</greatDelugeWaterLevelIncrementRatio>
+    </acceptor>
+    <forager>
+      <acceptedCountLimit>4</acceptedCountLimit>
```
Comment (triceo) [inline optaplanner-docs/src/main/asciidoc/LocalSearch/LocalSearch-chapter.adoc]: Didn't we change this to 1?

**Comment by triceo** [comment]
@BorisKocian Please take a look at the final set of comments. We will be able to merge this after that.

**Comment by ge0ffrey** [comment]
Hold on. Great deluge isn't ready for general consumption - it was still full of holes (see our last meeting with Boris), nothing I want to support in a backwards compatible manner.
Lukas, let's meet up today to discuss, but I think we need to roll this back or at least comment it out.

**Comment by ge0ffrey** [comment]
Meeting invite send for tomorrow morning.

**Pull request merged**